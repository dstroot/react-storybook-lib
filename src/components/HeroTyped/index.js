import React from "react";
import Typed from "react-typed";

const HeroTyped = ({ image, tagline, typed, strings }) => {
  const hero = {
    /* Image */
    backgroundImage: `url(${image})`,

    /* Set height and width */
    width: "100vw",
    height: "100vh",

    /* Position and center the image to scale nicely on all screens */
    backgroundPosition: "center",
    backgroundRepeat: "no-repeat",
    backgroundSize: "cover",
    position: "relative",

    /* create a shadow around the outside edge */
    boxShadow: "inset 0 0 10rem rgba(0, 0, 0, 1)"
  };

  // place the text on the hero image - center text, viewport wide
  const textSection = {
    textAlign: "center",
    position: "absolute",
    bottom: "25vh",
    width: "100vw",
    color: "white"
  };

  const textBig = {
    fontSize: `calc(52px + 1.5vw + 0.5vmin)`,
    marginBottom: "0",
    fontWeight: "500"
  };

  const textSmall = {
    marginTop: "-0.5vw",
    marginBottom: "0",
    fontSize: `calc(18px + 1.5vw + 0.5vmin)`,
    fontWeight: "300"
  };

  return (
    <div style={hero}>
      <div style={textSection}>
        <p style={textBig}>{tagline}</p>
        <p style={textSmall}>
          {typed}{" "}
          <Typed
            className="text-primary"
            strings={strings}
            typeSpeed={40}
            backSpeed={50}
            loop
          />
        </p>
        <a
          href="/"
          className="btn btn-lg btn-outline-primary mt-4"
          // onClick={ga(
          //   "send",
          //   "event",
          //   "Jumbotron actions",
          //   "Get started",
          //   "Get started"
          // )}
        >
          Get started!
        </a>
      </div>
    </div>
  );
};

export default HeroTyped;
