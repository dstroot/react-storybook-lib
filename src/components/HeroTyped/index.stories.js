import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, text, boolean, number } from "@storybook/addon-knobs";

import HeroTyped from "../HeroTyped";
import image from "../HeroTyped/media/hatch.jpg";

const stories = storiesOf("HeroTyped", module);
stories.addDecorator(withKnobs);

stories
  .addParameters({
    info: {
      inline: true,
      header: false,
      text: `
      This component requires the library "react-typed"

      ~~~js
      yarn add react-typed
      ~~~
      `
    }
  })
  .add("show example", () => (
    <HeroTyped
      image={image}
      tagline={text("Tagline", "Hatch Labs")}
      typed={text("Intro", "Hatch your ")}
      strings={["ideas", "innovations", "products", "solutions"]}
    />
  ));
