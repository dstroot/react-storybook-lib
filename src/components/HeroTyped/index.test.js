import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import HeroTyped from "../HeroTyped";
import image from "./media/hatch.jpg";

// What we care about is the functionality of <HeroTyped />, but not
// the dependent component "Typed" from "react-typed" (which breaks our tests!)
// So we will just mock it using jest so we can test the main functionality
// of our component.
jest.mock("react-typed", () => () => <span id="mockTyped">mockTyped</span>);

describe("HeroTyped", () => {
  // mount the component
  let mountedComponent;
  const getMountedComponent = () => {
    if (!mountedComponent) {
      mountedComponent = mount(
        <HeroTyped
          image={image}
          tagline="Hatch Labs"
          typed="Hatch your "
          strings={["ideas", "innovations", "products", "solutions"]}
        />
      );
    }
    return mountedComponent;
  };

  beforeEach(() => {
    mountedComponent = undefined;
  });

  it("it should render", () => {
    const component = renderer.create(
      <HeroTyped
        image={image}
        tagline="Hatch Labs"
        typed="Hatch your "
        strings={["ideas", "innovations", "products", "solutions"]}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("it should say Hatch Labs", () => {
    const p = getMountedComponent()
      .find("p")
      .first();

    expect(p.text()).toContain("Hatch Labs");
  });
});

// Testing information
// ---------------------------------------------------
// Jest - Expect: https://jestjs.io/docs/en/expect
// https://jestjs.io/docs/en/expect.html#content
// Enzyme cheatsheet - https://devhints.io/enzyme

// https://medium.com/@stipsan/testing-with-jest-15-awesome-tips-and-tricks-42150ec4c262
// https://www.youtube.com/watch?time_continue=459&v=XUlGzJLZe2Q
// https://medium.freecodecamp.org/the-right-way-to-test-react-components-548a4736ab22
// https://blog.pragmatists.com/click-your-element-how-to-make-enzyme-tests-cleaner-3e7ad21c2be0
// https://blog.bitsrc.io/how-to-test-react-components-with-jest-and-enzyme-in-depth-145fcd06b90
// https://medium.com/front-end-hacking/tested-react-lets-build-a-data-table-a76aa100d23f

// Component Contracts
// --------------------------------------------------
// In order to test a component you must first understand what its
// Contract is. Understanding a component’s contract is the most important
// part of testing a React component. A contract defines the expected behavior
// of your component and what assumptions are reasonable to have about its
// usage. Without a clear contract, your component may be hard to understand.
// Writing tests is a great way to formally define your component’s contract.

// Every React component has at least one thing that contributes to the
// definition of its contract:

// - What it renders (which may be nothing)
//
// Additionally, most component contracts are affected by these things as well:
//
// - The props the component receives
// - The state the component holds (if any)
// - What the component does when the user interacts with it
//   (via clicking, dragging, keyboard input, etc)
