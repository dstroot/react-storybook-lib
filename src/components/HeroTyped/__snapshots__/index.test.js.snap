// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroTyped it should render 1`] = `
<div
  style={
    Object {
      "backgroundImage": "url(hatch.jpg)",
      "backgroundPosition": "center",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "boxShadow": "inset 0 0 10rem rgba(0, 0, 0, 1)",
      "height": "100vh",
      "position": "relative",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "bottom": "25vh",
        "color": "white",
        "position": "absolute",
        "textAlign": "center",
        "width": "100vw",
      }
    }
  >
    <p
      style={
        Object {
          "fontSize": "calc(52px + 1.5vw + 0.5vmin)",
          "fontWeight": "500",
          "marginBottom": "0",
        }
      }
    >
      Hatch Labs
    </p>
    <p
      style={
        Object {
          "fontSize": "calc(18px + 1.5vw + 0.5vmin)",
          "fontWeight": "300",
          "marginBottom": "0",
          "marginTop": "-0.5vw",
        }
      }
    >
      Hatch your 
       
      <span
        id="mockTyped"
      >
        mockTyped
      </span>
    </p>
    <a
      className="btn btn-lg btn-outline-primary mt-4"
      href="/"
    >
      Get started!
    </a>
  </div>
</div>
`;
